tests/test_crawlers.py

```diff
@@ -599,3 +602,69 @@ def test_iterating(self):
         with self.assertLogs(self.crawler.LOGGER):
             for i, url in enumerate(self.crawler):
                 self.assertEqual(url, expected_urls[i])
+
+
+class FTPCrawlerTestCase(unittest.TestCase):
+    """Tests for the FTP crawler"""
+
+    def the_503_fake_ftp_exception(self, *args):
+        """Helper function for providing the correct ftp exception """
+        raise Exception("503")
+
+    def the_230_fake_ftp_exception(self, *args):
+        """Helper function for providing the correct ftp exception """
+        raise Exception("230")
+
+    def cwd_fake_function_of_ftp(self, name):
+        """passes in the case of "", ".." or "folder_name" in order to resemble the behavior of cwd
+        of ftplib. Otherwise (encountering a filename) raise the proper exception """
+        if (name == "..") or (name == "folder_name") or (name == ""):
+            pass
+        else:
+            raise ftplib.error_perm
+
+    @mock.patch('geospaas_harvesting.crawlers.ftplib.FTP', autospec=True)
+    def test_ftp_correct_navigation(self, mock_ftp):
+        """ shall categorize the specific file names (based on specific 'fileformat' which is
+        revealed in the configuration file) as well as folder(s) inside the ftp resource """
+        test_crawler = crawlers.FTPCrawler('ftp:///', fileformat='.gz')
+        test_crawler.ftp.nlst.return_value = ['file1.gz', 'folder_name', 'file3.bb', 'file2.gz', ]
+        test_crawler.ftp.cwd = self.cwd_fake_function_of_ftp
+        test_crawler.ftp.host = ''
+        with self.assertLogs('geospaas_harvesting.crawlers.FTPCrawler'):
+            test_crawler._explore_page('')
+        # '.gz' files must be in the "_urls" list
+        self.assertTrue(any([url_name.endswith('file1.gz') for url_name in test_crawler._urls]))
+        self.assertTrue(any([url_name.endswith('file2.gz') for url_name in test_crawler._urls]))
+        # Other type of files should not be in the "_urls" list
+        self.assertFalse(any([url_name.endswith('file3.bb') for url_name in test_crawler._urls]))
+        # folder with 'folder_name' must be in the "_to_process" list
+        self.assertTrue(any([url_name.endswith('folder_name')
+                             for url_name in test_crawler._to_process]))
+
+    @mock.patch('geospaas_harvesting.crawlers.ftplib.FTP.login')
+    def test_ftp_correct_exception(self, mock_ftp):
+        """ shall return 'AttributeError' instead of 'RuntimeError' in order to continuing
+        the harvesting process in the case of redundant or repetitive login attempt(s) after
+        the first login attempt """
+        test_crawler = crawlers.FTPCrawler('ftp:///', username="d", password="d", fileformat='.gz')
+        mock_ftp.side_effect = self.the_503_fake_ftp_exception
+        with self.assertRaises(AttributeError):
```

Comment:
Why does it raise AttributeError? And why is it good?

Reply:
@akorosov 
I have corrected it in order to be a crystal clear test. That previous version is because of reaching Nonetype in the cwd command of ftp. Now it has been corrected in the next commoit.